pyrogram/sync.py

```diff
@@ -40,49 +40,31 @@ def async_to_sync_wrap(*args, **kwargs):
         try:
             loop = asyncio.get_event_loop()
         except RuntimeError:
-            loop = main_loop
+            loop = asyncio.new_event_loop()
 
         if threading.current_thread() is threading.main_thread():
             if loop.is_running():
                 return coroutine
+            else:
+                if inspect.iscoroutine(coroutine):
+                    return loop.run_until_complete(coroutine)
 
-            if inspect.iscoroutine(coroutine):
-                return loop.run_until_complete(coroutine)
-
-            if inspect.isasyncgen(coroutine):
-                return loop.run_until_complete(consume_generator(coroutine))
-
-        elif loop is main_loop:
-            # This Thread hasn't an Event loop. So can do blocking calls
-            # the main_loop still may not be running
-            if inspect.iscoroutine(coroutine):
-                return asyncio.run_coroutine_threadsafe(coroutine, loop).result()
-
-            if inspect.isasyncgen(coroutine):
-                return asyncio.run_coroutine_threadsafe(consume_generator(coroutine), loop).result()
-
+                if inspect.isasyncgen(coroutine):
+                    return loop.run_until_complete(consume_generator(coroutine))
         else:
-            # This Thread has an Event loop. So better to do it in async way as possible
-            # the main_loop still may not be running
             if inspect.iscoroutine(coroutine):
-                if loop.is_running():
-                    # We can't just await future here.
-                    # Because it will instantly submit this coroutine to the main_loop.
-                    # Is there a better way ?
-                    async def coro_wrapper():
-                        return await asyncio.wrap_future(
-                            asyncio.run_coroutine_threadsafe(coroutine, main_loop))
+                future = asyncio.run_coroutine_threadsafe(coroutine, main_loop)
 
-                    return coro_wrapper()
-
-                return asyncio.run_coroutine_threadsafe(coroutine, main_loop).result()
+                if loop.is_running():
+                    return asyncio.wrap_future(future)
```

Comment:
in here, is it ok to return just the future ? shouldn't be wrapped with a coro ?
```python
async def coro():
    return await asyncio.wrap_future(asyncio.run_coroutine_threadsafe(coroutine, main_loop))

return coro()
```
because this will submit request to the main loop and it will be called soon inside main loop
```python
future = asyncio.run_coroutine_threadsafe(coroutine, main_loop)
return asyncio.wrap_future(future)
```
example:
```python
async def main():
    fut = client.get_me()
    # suppose we will need to drive this method later
    # but in here get_me() method will be called soon before we explicitly await

    ...

    me = await fut
    # actually we need to do it here right ?
```

Reply:
> in here, is it ok to return just the future ? shouldn't be wrapped with a coro ?

Thanks for noticing. Correct, it should indeed, like you proposed initially. `run_coroutine_threadsafe` actually does the scheduling part and immediately starts the task underneath and we want that only in case of non-async code.